apps/gateway/test/champions_test.exs

```diff
@@ -295,13 +280,134 @@ defmodule Gateway.Test.Champions do
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:campaigns, %Campaigns{} = campaigns}
+      }
+
+      assert Enum.count(campaigns.campaigns) > 0
+
+      sample_campaign = Enum.random(campaigns.campaigns)
+
+      # GetCampaign
+      SocketTester.get_campaign(socket_tester, user.id, sample_campaign.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:campaign, %Campaign{} = campaign_to_verify}
+      }
+
+      assert campaign_to_verify.id == sample_campaign.id
+
+      # GetLevel
+      level = Enum.random(sample_campaign.levels)
+
+      SocketTester.get_level(socket_tester, user.id, level.id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:level, %Level{} = level_to_verify}
+      }
+
+      assert level_to_verify.id == level.id
+    end
+
+    test "fight level", %{socket_tester: socket_tester} do
+      # Register user
+      {:ok, user} = Users.register("battle_user")
+
+      # Get user's first campaign progression
+      [campaign_progression | _] = user.campaign_progresses
+
+      # Get the level of the campaign progression
+      level_id = campaign_progression.level_id
+
+      # FightLevel
+      SocketTester.fight_level(socket_tester, user.id, level_id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:battle_result, _ = battle_result}
+      }
+
+      # Battle result should be either win or loss
+      assert battle_result.result == "win" or battle_result.result == "loss"
+
+      # TODO: check rewards [#CHoM-341]
+    end
+
+    test "fight level advances level in the campaign progression", %{socket_tester: socket_tester} do
+      # Register user
+      {:ok, user} = Users.register("battle_winning_user")
+
+      # Make user units very strong to win the battle
+      Enum.each(user.units, fn unit ->
+        GameBackend.Units.update_unit(unit, %{level: 9999})
+      end)
+
+      # Get user's first campaign progression
+      [campaign_progression | _] = user.campaign_progresses
+
+      # Get the level of the campaign progression
+      level_id = campaign_progression.level_id
+      level_number = campaign_progression.level.level_number
+
+      # FightLevel
+      SocketTester.fight_level(socket_tester, user.id, level_id)
+      fetch_last_message(socket_tester)
+
+      assert_receive %WebSocketResponse{
+        response_type: {:battle_result, _ = battle_result}
+      }
+
+      assert battle_result.result == "win"
+
+      {:ok, advanced_user} = Users.get_user(user.id)
+
+      [advanced_campaign_progression | _] = advanced_user.campaign_progresses
+
+      assert user.id == advanced_user.id
+      assert advanced_campaign_progression.level_id != level_id
+      assert advanced_campaign_progression.level.level_number > level_number
```

Comment:
if the progression will be always linear we should check that the number is `level_number + 1` to see that the campaign is progressing correctly